src/copy.rs

```diff
@@ -961,64 +994,74 @@ fn prepare_copy_internal(
-
-            match data_paths.entry(mime_type) {
+            match data_sources.entry(mime_type) {
                 Entry::Occupied(_) => {
-                    if cfg!(feature = "memfd") {
-                        data_path.pop();
-                    } else {
-                        // This MIME type has already been specified, so ignore it.
-                        remove_file(&*data_path).map_err(Error::TempFileRemove)?;
-                        data_path.pop();
-                        remove_dir(&*data_path).map_err(Error::TempDirRemove)?;
+                    // This MIME type has already been specified, so ignore it.
+                    match source {
+                        #[cfg(feature = "memfd")]
+                        FileSourceInner::InMemory(file) => drop(file),
+                        FileSourceInner::TempFile { file, mut path, .. } => {
+                            // Be explicit during cleanup.
+                            drop(file);
+                            // Get the tempdir parent created prior.
+                            path.pop();
+                            remove_dir_all(path).map_err(Error::TempFileRemove)?;
+                        }
                     }
                 }
                 Entry::Vacant(entry) => {
                     if !options.omit_additional_text_mime_types
-                        && text_data_path.is_none()
-                        && mime_type_is_text
+                        && text_data.is_none()
+                        && is_text(entry.key())
                     {
-                        text_data_path = Some(data_path.clone());
-                    }
-
-                    entry.insert(data_path);
-                    if cfg!(feature = "memfd") {
-                        state.memfd_files.push(file);
+                        text_data = Some((entry.into_key(), source));
+                    } else {
+                        entry.insert(source);
                     }
                 }
             }
         }
 
         // If the MIME type is text, offer it in some other common formats.
-        if let Some(text_data_path) = text_data_path {
+        if let Some((original_text_mime, text_data)) = text_data {
             let text_mimes = [
                 "text/plain;charset=utf-8",
                 "text/plain",
                 "STRING",
                 "UTF8_STRING",
                 "TEXT",
             ];
-            for &mime_type in &text_mimes {
-                // We don't want to overwrite an explicit mime type, because it might be bound to a
-                // different data_path
-                if !data_paths.contains_key(mime_type) {
-                    data_paths.insert(mime_type.to_string(), text_data_path.clone());
-                }
+
+            // XXX: The code below optimizes for minimal fd duplication/cloning since
+            // they are more likely to become exhausted or cause issues then small memory allocations.
```

Comment:
This is assuming the data you're copying is indeed small. Maybe we should choose one over the other depending on the size of the data? (i.e. if the data is large, duplicate the fd instead)

Reply:
To clarify further: The allocations here are not bounded on the size of the data currently stored/being served from the clipboard. The allocations are collections with always-known (small) data inside:
- `HashSet` of `&'static str`s
- `Vec` of `FileSourceInner`, which doesn't actually contain the copied data (only a FD that points to it).

In both the cases of the copied data being small or large we chose to duplicate the FDs, this code just tries to minimize how many FDs are opened by reusing the original.